### DISCUSSION TITLE
transcription using websocket and pyaudio

### DISCUSSION THREAD

**Discussion by Najaf-Zawar** [deepgram/community#707] [General help] [answered]
### Which Deepgram product feature are you using?

Deepgram API - STT Streaming

### Details

I am trying to input from microphone using pyaudio but its not doing transcription I am getting that error
data:  16000
data:  16000
data:  16000
chk1
on_close() takes 1 positional argument but 3 were given
data:  16000
WebSocket connection closed unexpectedly

### If you are making a request to the Deepgram API, what is the full Deepgram URL you are making a request to?

wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=16000&channels=1

### If possible, please attach your code or paste it into the text box.

import websocket
import base64
import pyaudio
import json
from threading import Thread
from api_secrets import API_KEY_DG

# YOUR_API_TOKEN = API_KEY_DG
FRAMES_PER_BUFFER = 8000
FORMAT = pyaudio.paInt16
CHANNELS = 1
SAMPLE_RATE = 16000
p = pyaudio.PyAudio()
 

# starts recording
stream = p.open(
   format=FORMAT,
   channels=CHANNELS,
   rate=SAMPLE_RATE,
   input=True,
   frames_per_buffer=FRAMES_PER_BUFFER
)

def on_message(ws, message):
    """
    is being called on every message
    """
    msg = json.loads(message)
    print('chk1')
    # print(msg)
    if 'channel' in msg and 'alternatives' in msg['channel'] and msg['channel']['alternatives']:
        transcript = msg['channel']['alternatives'][0]['transcript']
        print("chk2")
        if transcript:
            print(f'Transcript = {transcript}')
            # Pass transcript to play_audio function
            # play_audio(transcript)
                        

def on_error(ws, error):
    """
    is being called in case of errors
    """
    print(error)

def on_close(ws):
    """
    is being called on session end
    """
    print("WebSocket closed")

def on_open(ws):
    """
    is being called on session begin
    """
    def send_data():
        # while True:
            # read from the microphone
            # data = stream.read(FRAMES_PER_BUFFER, exception_on_overflow=False)
            
            # # encode the raw data into base64 to send it over the websocket
            # data = base64.b64encode(data).decode("utf-8")
            
            # # Follow the message format of the Real-Time service (see documentation)
            # json_data = json.dumps({"audio_data":str(data)})

            # # Send the data over the wire
            # ws.send(json_data)
            # print("data: ", len(data))
            # ws.send(data)
        try:
            while True:
                # read from the microphone
                data = stream.read(FRAMES_PER_BUFFER)
                print("data: ", len(data))
                ws.send(data)
        except websocket.WebSocketConnectionClosedException:
            print("WebSocket connection closed unexpectedly")
        except Exception as e:
            print("An error occurred:", e)
        finally:
            stream.stop_stream()
            stream.close()
            p.terminate()

    # Start a thread where we send data to avoid blocking the 'read' thread
    Thread(target=send_data).start()

# Set up the WebSocket connection with your desired callback functions
websocket.enableTrace(False)

# After opening the WebSocket connection, send an authentication header with your API token
extra_headers = {
        'Authorization': 'token ' + API_KEY_DG

    }

ws = websocket.WebSocketApp(
    f"wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=16000&channels=1",
    header=extra_headers,
    on_message=on_message,
    on_open=on_open,
    on_error=on_error,
    on_close=on_close
)

# Start the WebSocket connection
ws.run_forever()

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment, marked as answer]
Hi @Najaf-Zawar 

I am not exactly sure where you are getting the class `websocket.WebSocketApp` from or the `ws.run_forever()` (seems like you could be rolling your own websocket client instead of using the SDK), usually if you get a `WebSocket connection closed unexpectedly`, it's because either the audio stream isn't being sent OR the encoding option on the querystring doesn't match what the websocket endpoint is receiving.
